k2/csrc/pytorch_context.cu

```diff
@@ -149,18 +192,53 @@ class PytorchPinnedContext : public Context {
     return other.GetDeviceType() == kCpu;
   }
 
+  void CopyDataTo(size_t num_bytes, const void *src, ContextPtr dst_context,
+                  void *dst) override {
+    DeviceType device_type = dst_context->GetDeviceType();
+    switch (device_type) {
+      case kCpu:
+        // we assume that src and dst do not overlap
+        memcpy(dst, src, num_bytes);
+        break;
+      case kCuda: {
+        // from host to device
+        c10::cuda::CUDAStream stream = c10::cuda::getCurrentCUDAStream();
+        cudaError_t ret = cudaMemcpyAsync(dst, src, num_bytes,
+                                          cudaMemcpyHostToDevice, stream);
+        K2_CHECK_CUDA_ERROR(ret);
+
+        // CAUTION: we assume that src points to the beginning of the memory
+        // returned by `this` context. Otherwise, the following call is a no-op.
+        ret = THCCachingHostAllocator_recordEvent(const_cast<void *>(src),
+                                                  stream);
+        K2_CHECK_CUDA_ERROR(ret);
+        break;
+      }
+      default:
+        K2_LOG(FATAL) << "Unsupported device type: " << device_type;
+        break;
+    }
+  }
+
  private:
   torch::Allocator *allocator_;  // NOT owned here
 };
 
 ContextPtr GetCpuContext() { return std::make_shared<PytorchCpuContext>(); }
 
-ContextPtr GetPinnedContext() {
-  std::call_once(has_cuda_init_flag, InitHasCuda);
-
-  if (has_cuda) return std::make_shared<PytorchPinnedContext>();
-
-  return GetCpuContext();
+ContextPtr GetPinnedContext(DeviceType device_type) {
```

Comment:
I would rather have a function GetPinnedContext() that returns a pinned context, and a function
GetContextForTransfer(DeviceType) that returns the appropriate one.

Reply:
`GetContextForTransfer` is added but I don't know where it will be called.
I think we can always use `GetPinnedContext()`.

Reply:
No, we don't want to use pinned memory if we are doing CPU-only computations.  It's bad for the computer (stops it from swapping).. can make the system unstable if you use too much.  It's only for when you want to copy something to the device.

Reply:
I see, thanks.

Reply:
Fixed by adding `GetContextForTransfer`